wandb/integration/ultralytics/callback.py

```diff
@@ -0,0 +1,367 @@
+        checkpoint_dict = {
+            "epoch": trainer.epoch,
+            "best_fitness": trainer.best_fitness,
+            "model": copy.deepcopy(de_parallel(self.model)).half(),
+            "ema": copy.deepcopy(trainer.ema.ema).half(),
+            "updates": trainer.ema.updates,
+            "optimizer": trainer.optimizer.state_dict(),
+            "train_args": vars(trainer.args),
+            "date": datetime.now().isoformat(),
+            "version": __version__,
+        }
+        checkpoint_path = trainer.wdir / f"epoch{trainer.epoch}.pt"
+        torch.save(checkpoint_dict, checkpoint_path, pickle_module=pickle)
+        model_checkpoint_artifact.add_file(checkpoint_path)
+        wandb.log_artifact(
+            model_checkpoint_artifact, aliases=[f"epoch_{trainer.epoch}"]
+        )
+
+    def on_train_start(self, trainer: DetectionTrainer):
+        with telemetry.context(run=wandb.run) as tel:
+            tel.feature.ultralytics_yolov8 = True
+
+    def on_fit_epoch_end(self, trainer: DetectionTrainer):
+        validator = trainer.validator
+        dataloader = validator.dataloader
+        class_label_map = validator.names
+        with torch.no_grad():
+            self.device = next(trainer.model.parameters()).device
+            trainer.model.to("cpu")
+            self.model = copy.deepcopy(trainer.model).eval().to(self.device)
+            self.predictor.setup_model(model=self.model, verbose=False)
+            if isinstance(trainer, PoseTrainer):
+                self.train_validation_table = plot_pose_validation_results(
+                    dataloader=dataloader,
+                    class_label_map=class_label_map,
+                    predictor=self.predictor,
+                    visualize_skeleton=self.visualize_skeleton,
+                    table=self.train_validation_table,
+                    max_validation_batches=self.max_validation_batches,
+                    epoch=trainer.epoch,
+                )
+            elif isinstance(trainer, DetectionTrainer):
+                self.train_validation_table = plot_validation_results(
+                    dataloader=dataloader,
+                    class_label_map=class_label_map,
+                    predictor=self.predictor,
+                    table=self.train_validation_table,
+                    max_validation_batches=self.max_validation_batches,
+                    epoch=trainer.epoch,
+                )
+            elif isinstance(trainer, ClassificationTrainer):
+                self.train_validation_table = plot_classification_validation_results(
+                    dataloader=dataloader,
+                    predictor=self.predictor,
+                    table=self.train_validation_table,
+                    max_validation_batches=self.max_validation_batches,
+                    epoch=trainer.epoch,
+                )
+        if self.enable_model_checkpointing:
+            self._save_model(trainer)
+        trainer.model.to(self.device)
```

Comment:
Maybe also move the `self.model` to `cpu` or `del self.model` and run `gc collect()`. Because it looks like after the first epoch `self.model` will always be on `device` otherwise.